include/deal.II/lac/la_parallel_vector.templates.h

```diff
@@ -629,79 +631,43 @@ namespace LinearAlgebra
     Vector<Number, MemorySpaceType>::
     operator=(const Vector<Number, MemorySpaceType> &c)
     {
-#ifdef _MSC_VER
-      return this->operator=<Number>(c);
-#else
-      return this->template operator=<Number>(c);
-#endif
+      // we update ghost values whenever one of the input or output vector
+      // already held ghost values or when we import data from a vector with
+      // the same local range but different ghost layout
+      assignment(c, c.vector_is_ghosted);
+
+      return *this;
     }
 
 
 
     template <typename Number, typename MemorySpaceType>
-    template <typename Number2>
+    template <typename MemorySpaceType2>
     inline Vector<Number, MemorySpaceType> &
     Vector<Number, MemorySpaceType>::
-    operator=(const Vector<Number2, MemorySpaceType> &c)
+    operator=(const Vector<Number, MemorySpaceType2> &c)
     {
-      Assert(c.partitioner.get() != nullptr, ExcNotInitialized());
-
       // we update ghost values whenever one of the input or output vector
       // already held ghost values or when we import data from a vector with
       // the same local range but different ghost layout
-      bool must_update_ghost_values = c.vector_is_ghosted;
+      assignment(c, c.vector_is_ghosted);
```

Comment:
I don't think you can pass in `c.vector_is_ghosted` as an argument because we do not want to change the ghosted value of the other constant vector `c` (you can only change it because the variable is `mutable`). We should create a separate value. Ideally, we should simply return the `is_ghosted` information from the `assignment` function and resolve the boolean variable inside.